File: assignment4/monte2.cpp
#include <iostream>
#include <cstdio>
#include <cmath>

#define LL long long

using namespace std;

double func_rand(LL int val, LL int m){
	return (double)val/m;
}

LL int func_val(LL int val, LL int a, LL int b, LL int m){
	LL next_val;
	next_val = ((a * val) + b) % m;

	return next_val;
}

void function(LL int initial, LL int a, LL int b, LL int m, int end, double array[]){
	LL int val, count = 0;
	LL int index, quot;
	double random, prev;

	val = initial;
	random = func_rand(val, m);

	for(count=0;count<end;++count){
		val = func_val(val, a, b, m);
		random = func_rand(val, m);
		
		array[count] = random;
	}
}

void exponential(double lambda, double array[], LL int bound){
	int index;
	
	for(index=0; index<bound; ++index){
		array[index] = -(double)log(1.0 - array[index])/lambda;
	}
}

int main(){
	LL int index;
	double array1[5005];
	double array2[5005];
	double array3[5005];
	double array4[5005];
	double array5[5005];
	double array[5005];
	LL int init_value[5] = {3452, 997, 34, 269, 5};
	LL int bound = 5000;
	double lambda = 5;
	double min, max, mean, sum = 0.0;

	function(init_value[0], 4696, 0, 5003, bound, array1);
	exponential(lambda, array1, bound);
	function(init_value[1], 4696, 0, 5003, bound, array2);
	exponential(lambda, array2, bound);
	function(init_value[2], 4696, 0, 5003, bound, array3);
	exponential(lambda, array3, bound);
	function(init_value[3], 4696, 0, 5003, bound, array4);
	exponential(lambda, array4, bound);
	function(init_value[4], 4696, 0, 5003, bound, array5);
	exponential(lambda, array5, bound);

	for(index=0; index<bound; ++index){
		array[index] = array1[index] + array2[index] + array3[index] + array4[index] + array5[index];
		printf("%lf\n", array[index]);
	}

	min = max = array[0];
	for(index=0; index<bound; ++index){
		sum+=array[index];
		
		if(array[index] < min)
			min = array[index];
			
		if(array[index] > max)
			max = array[index];
	}
	mean = (double)sum/bound;
	
	printf("Min=%lf\nMax=%lf\nMean=%lf\n", min, max, mean);
	
	return 0;
}

File: assignment5/monte2.R
func_rand <- function(val, m){
	return(val/m);
}

func_val <- function(val, a, b, m){
	next_val <- ((a * val) + b) %% m;
	return(next_val);
}

func <- function(a, b, m, end, lambda){
	arr1 <- array(end);
	arr2 <- array(end);
	arr3 <- array(end);
	arr4 <- array(end);
	arr5 <- array(end);
	arr <- array(end);

	val1 <- 3452;
	val2 <- 997;
	val3 <- 34;
	val4 <- 269;
	val5 <- 5;

	for(count in 1:end){
		val1 <- func_val(val1, a, b, m);
		arr1[count] <- func_rand(val1, m);
		arr1[count] <- -log(1 - arr1[count])/lambda;
		
		val2 <- func_val(val2, a, b, m);
		arr2[count] <- func_rand(val2, m);
		arr2[count] <- -log(1 - arr2[count])/lambda;
		
		val3 <- func_val(val3, a, b, m);
		arr3[count] <- func_rand(val3, m);
		arr3[count] <- -log(1 - arr3[count])/lambda;
		
		val4 <- func_val(val4, a, b, m);
		arr4[count] <- func_rand(val4, m);
		arr4[count] <- -log(1 - arr4[count])/lambda;
		
		val5 <- func_val(val5, a, b, m);
		arr5[count] <- func_rand(val5, m);
		arr5[count] <- -log(1 - arr5[count])/lambda;
		
		val <- arr1[count] + arr2[count] + arr3[count] + arr4[count] + arr5[count];
		arr[count] <- val;
		print(arr[count])
	}
	
	sum <- 0.0;
	min <- arr[1];
	max <- arr[1];
	
	for(index in 1:end){
		sum <- sum + arr[index];
		
		if(arr[index] < min){
			min <- arr1[index];
		}
			
		if(arr[index] > max){
			max <- arr1[index];
		}
	}
	
	mean <- sum/end;

	cat(sprintf("Min=%g\nMax=%g\nMean=%g\n", min, max, mean));

	hist(arr, 100,freq = TRUE, include.lowest = TRUE, right = TRUE, density = NULL,
		angle = 45, col = '#1E90FF', border = NULL,
		main = paste("Histogram of frequencies and total values
		generated=",end), xlim=range(0,3.5), ylim=range(0,(300)), xlab = 'Random Values' ,
		axes = TRUE, plot = TRUE, labels = FALSE);

	dev.copy(jpeg,"hist1.jpg");
	dev.off();	
}

main<-function(){
	bound <- 5000;
	lambda <- 5;

	func(4696, 0, 5003, bound, lambda);
}
